private/pkg/storage/filter.go

```diff
@@ -88,16 +88,22 @@ func (r *filterReadBucketCloser) Walk(ctx context.Context, prefix string, f func
 	if err != nil {
 		return err
 	}
-	return r.delegate.Walk(
-		ctx,
-		prefix,
-		func(objectInfo ObjectInfo) error {
-			if !r.matcher.MatchPath(objectInfo.Path()) {
-				return nil
-			}
-			return f(objectInfo)
-		},
-	)
+	prefixRoots := r.matcher.MatchPrefix(prefix)
+	for _, prefixRoot := range prefixRoots {
+		if err := r.delegate.Walk(
+			ctx,
+			prefixRoot,
+			func(objectInfo ObjectInfo) error {
+				if !r.matcher.MatchPath(objectInfo.Path()) {
+					return nil
```

Comment:
Ideally we would be able to skip subdirectories if the prefix is excluded as this Walk will still iterate over excluded child paths.